Include BuildCommunity in built type checks

avatus opened the pull request:
This commit will add two new methods to the modules interface, IsEnterprise and IsOSS. This is to help alieviate any pain around the code base that used to look specifically for BuildOSS. Now that we have two "oss types" (oss and community), we can use these methods to do enterprise/oss checks, rather than specifically looking at the build type. This also will allow any future build types to not affect these checks again in the future.

`e` counterpart: https://github.com/gravitational/teleport.e/pull/4316

codingllama commented on lib/modules/modules.go:
```diff
@@ -379,6 +383,16 @@ func (p *defaultModules) BuildType() string {
 }
 
+// IsEnterprise returns if the binary was built with enterprise modules
+func (p *defaultModules) IsEnterprise() bool {
```
Maybe IsEnterpriseBuild or IsBuildEnterprise?

Same below.

avatus replied:
sure, i'll do enterprise build

codingllama reviewed (commented):
Many thanks!

codingllama commented:
Friendly reminder that we need https://github.com/gravitational/teleport.e/pull/4316 to land before this one, so we can avoid build breakages.

avatus commented:
@camscale @fheinecke it looks like the `tsh_test` commands were returning a build type of `CLI`, so I assumed that any previous `== BuildOSS` check already failed so I opted to make the `cliModules` return false for both oss and enterprise. does that sound about right?

codingllama commented on lib/modules/modules.go:
```diff
@@ -379,6 +383,16 @@ func (p *defaultModules) BuildType() string {
 	return teleportBuildType
 }
 
+// IsEnterpriseBuild returns if the binary was built with enterprise modules
```
```suggestion
// IsEnterpriseBuild returns false for [defaultModules].
```

Or ditch the comment, it's a private type anyway. Similarly for others.

codingllama commented on lib/modules/modules.go:
```diff
@@ -379,6 +383,16 @@ func (p *defaultModules) BuildType() string {
+	return false
+}
+
+// IsOSSBuild returns if the binary was built without enterprise modules
```
```suggestion
// IsOSSBuild returns true for [defaultModules].
```

codingllama commented on tool/teleport/testenv/test_server.go:
```diff
@@ -393,6 +393,16 @@ func (p *cliModules) BuildType() string {
 	return "CLI"
 }
 
+// IsEnterpriseBuild returns if the binary was built with enterprise modules
```
```suggestion
// IsEnterpriseBuild returns false for [cliModules].
```

codingllama commented on tool/teleport/testenv/test_server.go:
```diff
@@ -393,6 +393,16 @@ func (p *cliModules) BuildType() string {
+	return false
+}
+
+// IsOSSBuild returns if the binary was built without enterprise modules
```
```suggestion
// IsOSSBuild returns false for [cliModules].
```

It is kinda of weird that this is neither. 🤷

codingllama commented on tool/tsh/common/tsh_test.go:
```diff
@@ -208,6 +208,16 @@ func (p *cliModules) BuildType() string {
 	return "CLI"
 }
 
+// IsEnterpriseBuild returns if the binary was built with enterprise modules
```
```suggestion
// IsEnterpriseBuild returns false for [cliModules].
```

codingllama commented on tool/tsh/common/tsh_test.go:
```diff
@@ -208,6 +208,16 @@ func (p *cliModules) BuildType() string {
+	return false
+}
+
+// IsOSSBuild returns if the binary was built without enterprise modules
```
```suggestion
// IsOSSBuild returns false for [cliModules].
```

codingllama reviewed (commented):
Looks good, but let's the enterprise PR first and pull it here.

codingllama commented on lib/modules/modules.go:
```diff
@@ -293,6 +293,10 @@ type Modules interface {
 	SetFeatures(Features)
 	// BuildType returns build type (OSS, Community or Enterprise)
 	BuildType() string
+	// IsEnterpriseBuild returns false for [defaultModules].
+	IsEnterpriseBuild() bool
+	// IsOSSBuild returns true for [defaultModules].
```
I think you changed a little too much here, these were fine.

avatus replied:
Oh, you're right. this is the interface. will revert, thanks

avatus commented:
Once the e pr is in, I'll add the e ref to this and squash

avatus committed:
Include BuildCommunity in built type checks

This commit will add two new methods to the modules interface,
IsEnterprise and IsOSS. This is to help alieviate any pain around the
code base that used to look specifically for BuildOSS. Now that we have
two "oss types" (oss and community), we can use these methods to do
enterprise/oss checks, rather than specifically looking at the build
type. This also will allow any future build types to not affect these
checks again in the future.

avatus committed:
Update e

codingllama reviewed (approved):
Many thanks!
